sharding/xorValidatorShuffler_test.go

```diff
@@ -0,0 +1,778 @@
+	nbLeavingPerShard := make(map[uint32]int)
+
+	nbLeavingFromEligible := 0
+	for i, eligibleList := range eligible {
+		nbWantingToLeaveFromList := numberMatchingNodes(eligibleList, prevLeaving)
+		maxAllowedToLeaveFromList := len(waiting[i])
+		nbLeaving := nbWantingToLeaveFromList
+		if nbLeaving > maxAllowedToLeaveFromList {
+			nbLeaving = maxAllowedToLeaveFromList
+		}
+
+		nbLeavingPerShard[i] += nbLeaving
+		nbLeavingFromEligible += nbLeaving
+	}
+	assert.Equal(t, nbLeavingFromEligible, len(prevLeaving)-len(newLeaving))
+
+	return nbLeavingFromEligible, nbLeavingPerShard
+}
+
+func testShuffledOut(
+	t *testing.T,
+	eligibleMap map[uint32][]Validator,
+	waitingMap map[uint32][]Validator,
+	newEligible map[uint32][]Validator,
+	shuffledOut []Validator,
+	prevleaving []Validator,
+	newleaving []Validator,
+) {
+	nbAllLeaving, _ := testLeaving(t, eligibleMap, waitingMap, prevleaving, newleaving)
+	allWaiting := getValidatorsInMap(waitingMap)
+	allEligible := getValidatorsInMap(eligibleMap)
+	assert.Equal(t, len(shuffledOut)+nbAllLeaving, len(allWaiting))
+
+	allNewEligible := getValidatorsInMap(newEligible)
+	assert.Equal(t, len(allEligible)-len(shuffledOut)-nbAllLeaving, len(allNewEligible))
+
+	newNodes := append(allNewEligible, shuffledOut...)
+	assert.NotEqual(t, allEligible, newNodes)
+	assert.True(t, contains(newNodes, allEligible))
+}
+
+func createDefaultXorShuffler() *randXORShuffler {
+	return NewXorValidatorsShuffler(
+		100,
+		100,
+		0.2,
+		false,
+	)
+}
+
+func getValidatorsInMap(valMap map[uint32][]Validator) []Validator {
+	result := make([]Validator, 0)
+
+	for _, valList := range valMap {
+		result = append(result, valList...)
+	}
+
+	return result
+}
+
+func Test_xorBytes_SameLen(t *testing.T) {
+	t.Parallel()
+
+	rez := xorBytes(firstArray, secondArray)
+
+	assert.Equal(t, expectedArray, rez)
+}
+
+func Test_xorBytes_FirstLowerLen(t *testing.T) {
+	t.Parallel()
+
+	rez := xorBytes(firstArray[:len(firstArray)-1], secondArray)
+
+	assert.Equal(t, expectedArray[:len(expectedArray)-1], rez)
+}
+
+func Test_xorBytes_SecondLowerLen(t *testing.T) {
+	t.Parallel()
+
+	rez := xorBytes(firstArray, secondArray[:len(secondArray)-1])
+
+	assert.Equal(t, expectedArray[:len(expectedArray)-1], rez)
+}
+
+func Test_xorBytes_FirstEmpty(t *testing.T) {
+	t.Parallel()
+
+	rez := xorBytes([]byte{}, secondArray)
+
+	assert.Equal(t, []byte{}, rez)
+}
+
+func Test_xorBytes_SecondEmpty(t *testing.T) {
+	rez := xorBytes(firstArray, []byte{})
+
+	assert.Equal(t, []byte{}, rez)
+}
+
+func Test_xorBytes_FirstNil(t *testing.T) {
+	t.Parallel()
+
+	rez := xorBytes(nil, secondArray)
+
+	assert.Equal(t, []byte{}, rez)
+}
+
+func Test_xorBytes_SecondNil(t *testing.T) {
+	t.Parallel()
+
+	rez := xorBytes(firstArray, nil)
+
+	assert.Equal(t, []byte{}, rez)
+}
+
+func Test_copyValidatorMap(t *testing.T) {
+	t.Parallel()
+
+	valMap := generateValidatorMap(30, 2)
+	v2 := copyValidatorMap(valMap)
+	assert.Equal(t, valMap, v2)
+
+	valMap[0] = valMap[0][1:]
```

Comment:
this can be replaced with `assert.False(t, valmap == v2)` as it tests pointer values

Reply:
I wanted to check deep equal not pointer value, copyValidatorMap does a deep copy.

Reply:
ok